### ISSUE TITLE
Substitute environment variables in clusterctl.yaml

### ISSUE THREAD

**Issue by omniproc** [kubernetes-sigs/cluster-api#5457]
**User Story**

As a operator I would like to use variables inside my `clusterctl.yaml` that will be substrituted with environment variables so we can use relative paths inside `clusterctl.yaml` e.g. to configure relative, local URIs.

**Detailed Description**

In an air-gapped environment you'll often find yourself in a situation where you have to provide all images and manifests localy when initializing a CAPI cluster. To have the setup documented and reproduceable it makes sense to package all the dependencies on your local CRD manifest files as well as the used `clusterctl.yaml` in some source code management system like Git.
Because the content of `clusterctl.yaml` is currently static it's not possible however to make such a setup independet of the system we're executing `clusterctl` from since e.g. we can't use relative paths or other system specific configurations. You end up having to edit the `clusterctl.yaml` depending on which system you're executing `clusterctl` on.

I'd like to request a new feature to change the YAML processor used to read the `clusterctl.yaml`  so it has an option to substitute variables, e.g.
in this example:

```YAML
providers:
- name: "cluster-api"
  type: "CoreProvider"
  url: "$HOME/.cluster-api/overrides/cluster-api/v1.0.0/core-components.yaml"
```

the `url` would be expanded to `/home/myuser/.cluster-api/overrides/cluster-api/v1.0.0/core-components.yaml`. This is just a single example on how this could be used to make the use of `clusterctl.yaml` more flexible.
I'm not sure which YAML processor is used currently for this but processors exist that provide such functions already as well as the required escape functions in case someone actually would like to use the string `$HOME` and not have that substituted in the example above.

/kind feature

**Comment by omniproc** [comment]
I'm good with envsubst and think it's syntax probably more reliable then bash. The bash syntax like example I gave above was just for illustration.

**Comment by Jont828** [comment]
/assign
@sbueringer @fabriziopandini I'd like to give this a shot. Where would we want to handle the variable substitution? Would we want to make changes to the Go code where the config file is read in, i.e. in [this file](https://github.com/kubernetes-sigs/cluster-api/blob/56d99d7bca51cd1506b24c6c7dc7cc4a6f458576/cmd/clusterctl/cmd/root.go#L113)? It looks like we handle stuff for envsubst in the Makefile as well so I'm not sure what the best way too approach the problem is.

**Comment by fabriziopandini** [comment]
@ykakarap to provide guidance

**Comment by sbueringer** [comment]
@Jont828 Sorry missed this in Feburary. I've implemented a solution for it in #7343 (we would like to use the functionality at a KubeCon tutorial in two weeks)

/assign
